spec/execution_tests.lua

```diff
@@ -1749,6 +1749,71 @@ function execution_tests.run(compile_file, backend, _ENV, only_compile)
         end)
     end)
 
+    describe("For-in loops", function()
+        compile([[
+            local function iter(state: any, current: any): (any, any)
+                if current as integer == #(state as {any}) then
+                    return nil, nil
+                end
+
+                return current as integer + 1, (state as {any})[current as integer + 1]
+            end
+
+            typealias iterfn = (any, any) -> (any, any)
+            local function xpairs(xs: {any}): (iterfn, any, any)
+                return iter, xs, 0
+            end
+
```

Comment:
I think we can improve this a bit by taking more advantage of implicit casting and by renaming the function to make it clear that it is reimplementing ipairs. But still, thanks a lot for the example you gave. It's easier to suggest an improvement than to come up with it from scratch ;)

```
local function iter(arr: {any}, prev: integer): (any, any)
    local i = prev + 1
    local x = arr[i]
    if x == (nil as any) then
        return nil, nil
    else
        return i, x
    end
end

typealias iterfn = (any, any) -> (any, any)
export function my_ipairs(xs: {any}): (iterfn, any, any)
    return iter, xs, 0
end
```